### PR TITLE
Implement missing protocols

### LINKED ISSUE #97: Implement `/ipcidr`
Add support for `/ipcidr`.

See specification added with https://github.com/multiformats/multiaddr/pull/129.

Motivation is to specify address ranges to (a) not announce or (b) not dial in https://github.com/mxinden/rust-libp2p-server/ used as an IPFS bootstrap node.

### PR DESCRIPTION
**Pull request by Zollerboy1**
This PR implements all protocols listed in https://github.com/multiformats/multiaddr/blob/master/protocols.csv.

Presumably, this PR closes #97.

### CODE REVIEW

**Comment by umgefahren** [comment]
I would like to see that merged especially since we need ip6zone

**Comment by umgefahren** [comment]
What's the status on this? @jxs

**Comment by Zollerboy1** [comment]
Hey @jxs, of course, I can do that. I'm on vacation right now, so I probably cannot work on this until next week though.

**Comment by Zollerboy1** [comment]
Hey @jxs, I finally came around to do this, it should be much easier to review now.

I have the code with the reordered protocols in another branch and would create a new PR when (or if) this one gets merged.

Currently, the CI is failing because the new `rust-check-cfg` feature in Rust 1.80 complains about the use of `#[cfg(nightly)]` in the test file. I wasn't sure how exactly you'd want to solve this, but this should probably be changed to `#[cfg(feature = "nightly")]` or similar.

**Comment by umgefahren** [comment]
> Currently, the CI is failing because the new `rust-check-cfg` feature in Rust 1.80 complains about the use of `#[cfg(nightly)]` in the test file. I wasn't sure how exactly you'd want to solve this, but this should probably be changed to `#[cfg(feature = "nightly")]` or similar.

I don't think adding an extra feature that only pertains to a test is a good idea. Using this cfg is actually not that uncommon. I would suggest you add the lint exception:

```toml
[lints.rust]
unexpected_cfgs = { level = "warn", check-cfg = ['cfg(nightly)'] }
```

It might seem like a dirty fix, but it's a widely used workaround:
- [`cfg(nightly)` exception in Cargo.toml](https://github.com/search?q=cfg%28nightly%29+lang%3Atoml&type=code)
- [`cfg(nightly)` in Rust files](https://github.com/search?q=cfg%28nightly%29+lang%3Arust+&type=code)

**Comment by umgefahren** [comment]
I made a quick PR: #112

**Comment by jxs** [comment]
Hi @Zollerboy1 can you pull the latest changes so we can make CI pass?
Thanks!

**Comment by umgefahren** [comment]
> Hi @Zollerboy1 can you pull the latest changes so we can make CI pass? Thanks!

It passes now.

File: src/protocol.rs [hunk 1]
```diff
@@ -224,6 +239,44 @@ impl<'a> Protocol<'a> {
...
+            "ipcidr" => {
+                let s = iter.next().ok_or(Error::InvalidProtocolString)?;
+                Ok(Protocol::Ipcidr(s.parse()?))
+            }
+            "garlic64" => {
+                let s = iter
+                    .next()
+                    .ok_or(Error::InvalidProtocolString)?
+                    .replace('-', "+")
+                    .replace('~', "/");
```
Comment (jxs) [inline src/protocol.rs:255]: why do we need replacing these characters?

**Review by jxs — COMMENTED** [review]
Hi! And sorry for the delay overal LGTM, just one question

File: src/protocol.rs (hunk 1 above)
Comment (Zollerboy1) [reply src/protocol.rs:255]: The garlic protocols are very weird in that there is almost no documentation of them anywhere. But apparently, garlic64 doesn't use classic Base64 encoding, but a similar encoding where `+` is replaced with `-` and `/` with `~`. I don't know why. But to use `multibase::Base::Base64.decode` on the string, we have to have a proper Base64 encoding, so I'm just replacing these characters. You can see in the multiaddr implementations in Go (https://github.com/multiformats/go-multiaddr/blob/414c602b9282d609668bd58c90c800468816d9f4/transcoders.go#L270-L271) and Java (https://github.com/multiformats/java-multiaddr/blob/ffdc5320b4ccb5d5e490b6bce72a8da5dbe775ae/src/main/java/io/ipfs/multiaddr/Protocol.java#L231) that they're doing the same.

Comment (jxs) [reply src/protocol.rs:255]: thanks for taking the time to explain and provide examples!

**Review by jxs — APPROVED** [review]
LGTM

**Pull request merged**